libnullpay/src/payment_method.rs

```diff
@@ -1,227 +1,305 @@
-                    return (cb)(cmd_handle, err, res.as_ptr());
-                }
-            } else {
-                panic!("Can't lock injections mutex");
-            }
-
-            handle(cmd_handle, $first_param_name, $($param_name,)* cb)
-        }
-
-        pub fn inject_mock(err: ErrorCode, res: *const c_char) {
-            if let Ok(mut injections) = INJECTIONS.lock() {
-                injections.push_back((err, CString::from(unsafe { CStr::from_ptr(res) })))
-            } else {
-                panic!("Can't lock injections mutex");
-            }
-        }
-
-        pub fn clear_mocks() {
-            if let Ok(mut injections) = INJECTIONS.lock() {
-                injections.clear();
-            } else {
-                panic!("Can't lock injections mutex");
-            }
-        }
-    )
-}
 
 pub static PAYMENT_METHOD_NAME: &str = "null";
 
 pub mod create_payment_address {
     use super::*;
 
-    mocked_handler!(wallet_handle: i32, config: *const c_char);
-
-    fn handle(cmd_handle: i32, _wallet_handle: i32, _config: *const c_char, cb: IndyPaymentCallback) -> ErrorCode {
-        let res = CString::new(format!("pay:null:{}", rand::get_rand_string(15))).unwrap();
+    pub extern fn handle(cmd_handle: i32, _wallet_handle: i32, _config: *const c_char, cb: Option<IndyPaymentCallback>) -> ErrorCode {
+        let res = format!("pay:null:{}", rand::get_rand_string(15));
         let err = ErrorCode::Success;
-        (cb)(cmd_handle, err, res.as_ptr())
+        _process_callback(cmd_handle, err, res, cb)
     }
 }
 
 pub mod add_request_fees {
     use super::*;
 
-    mocked_handler!(wallet_handle: i32, submitter_did: *const c_char, req_json: *const c_char, inputs_json: *const c_char, outputs_json: *const c_char);
+    pub extern fn handle(cmd_handle: i32, _wallet_handle: i32, _submitter_did: *const c_char, req_json: *const c_char, inputs_json: *const c_char, outputs_json: *const c_char, cb: Option<IndyPaymentCallback>) -> ErrorCode {
+        let res = unsafe { CStr::from_ptr(req_json).to_str() }.unwrap();
+        let inputs_json = unsafe { CStr::from_ptr(inputs_json).to_str() }.unwrap();
+        let outputs_json = unsafe { CStr::from_ptr(outputs_json).to_str() }.unwrap();
 
-    fn handle(cmd_handle: i32, _wallet_handle: i32, _submitter_did: *const c_char, req_json: *const c_char, _inputs_json: *const c_char, _outputs_json: *const c_char, cb: IndyPaymentCallback) -> ErrorCode {
-        let res = req_json;
-        let err = ErrorCode::Success;
-        (cb)(cmd_handle, err, res)
+        let inputs = from_str::<Vec<String>>(inputs_json);
+        let outputs = from_str::<Vec<UTXOOutput>>(outputs_json);
+
+        let (inputs, outputs) = match (inputs, outputs) {
+            (Ok(inputs), Ok(outputs)) => (inputs, outputs),
+            _ => { return ErrorCode::CommonInvalidStructure; }
+        };
+
+        let txn_type = match parse_operation_from_request(res) {
```

Comment:
Can we just use "?" operator? A lot of matches like this doesn't look really nice